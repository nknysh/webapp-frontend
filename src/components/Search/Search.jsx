import React, { Fragment, useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { __, lensProp, defaultTo, set, view, pipe, prop } from 'ramda';
import { renameKeys } from 'ramda-adjunct';
import debounce from 'lodash.debounce';
import { DatePicker, Checkbox, Button } from '@pure-escapes/webapp-ui-components';

import LodgingSelect from 'components/LodgingSelect';
import { isLoading } from 'store/common';
import { toDate } from 'utils';

import { propTypes, defaultProps } from './Search.props';
import { SearchBarIndexSearch, SearchBarSection } from './Search.styles';

const destinationLens = lensProp('destination');
const datesLens = lensProp('dates');
const lodgingLens = lensProp('lodging');
const repeatGuestLens = lensProp('repeatGuest');

export const getMonthToDisplay = pipe(
  prop('startDate'),
  toDate
);

const renderSearchButton = (t, { canSearch, onSubmit }) => (
  <Button disabled={!canSearch} onClick={onSubmit}>
    {t('buttons.search')}
  </Button>
);

export const Search = ({
  canSearch,
  onChange,
  onSearch,
  onSubmit,
  searchQuery,
  searchStatus,
  showSubmit,
  vertical,
  nameSearchResults,
}) => {
  const { t } = useTranslation();

  const updateSearchQuery = useCallback(set(__, __, searchQuery), [searchQuery]);
  const getSearchQueryData = useCallback(view(__, searchQuery), [searchQuery]);

  const setSelectedDatesToSearchQuery = useCallback(
    pipe(
      renameKeys({ from: 'startDate', to: 'endDate' }),
      updateSearchQuery(datesLens),
      onChange
    ),
    [updateSearchQuery, onChange]
  );

  const setLodgingsToSearchQuery = useCallback(
    pipe(
      updateSearchQuery(lodgingLens),
      onChange
    ),
    [updateSearchQuery, onChange]
  );

  const setRepeatGuest = useCallback(
    pipe(
      updateSearchQuery(repeatGuestLens),
      onChange
    ),
    [updateSearchQuery, onChange]
  );

  const onIndexSelect = useCallback(
    result => {
      // we need to do specific logic if the user has chosen to clean,
      // because actually searching is heavily tied into setting the search query
      // @see https://pureescapes.atlassian.net/browse/OWA-628
      if (result.id === 'CLEAR') {
        return pipe(onChange);
      } else {
        return pipe(
          updateSearchQuery(destinationLens),
          onChange
        );
      }
    },
    [updateSearchQuery, onChange]
  );

  const onRepeatGuestChange = useCallback((e, checked) => setRepeatGuest(checked), [setRepeatGuest]);

  // index an `admins` option into the select
  // @see https://pureescapes.atlassian.net/browse/OWA-628
  nameSearchResults['admins'] = [
    {
      id: 'CLEAR',
      name: t('form.placeholders.nameSearchClear'),
    },
  ];

  return (
    <Fragment>
      <SearchBarSection data-vertical={vertical}>
        <SearchBarIndexSearch
          isLoading={isLoading(searchStatus)}
          label={t('labels.search')}
          onChange={debounce(onSearch, 600)}
          onClick={onIndexSelect}
          placeholder={t('form.placeholders.search')}
          results={nameSearchResults}
          value={prop('value', getSearchQueryData(destinationLens))}
        />
      </SearchBarSection>
      <SearchBarSection data-vertical={vertical} data-large={true}>
        <DatePicker
          dayPickerProps={{
            month: getMonthToDisplay(getSearchQueryData(datesLens)),
          }}
          label={`${t('labels.dates')} *`}
          onSelected={setSelectedDatesToSearchQuery}
          selectedValues={getSearchQueryData(datesLens)}
        />
      </SearchBarSection>
      <SearchBarSection data-vertical={vertical}>
        <LodgingSelect
          label={`${t('labels.lodging')} *`}
          onSelected={setLodgingsToSearchQuery}
          rooms={getSearchQueryData(lodgingLens)}
        />
      </SearchBarSection>
      <SearchBarSection data-vertical={vertical} data-constrain={true}>
        <Checkbox
          label={t('labels.isRepeat')}
          onChange={onRepeatGuestChange}
          checked={defaultTo(false, getSearchQueryData(repeatGuestLens))}
        />
      </SearchBarSection>

      {showSubmit && (
        <SearchBarSection data-vertical={vertical}>{renderSearchButton(t, { onSubmit, canSearch })}</SearchBarSection>
      )}
    </Fragment>
  );
};

Search.propTypes = propTypes;
Search.defaultProps = defaultProps;

export default Search;
