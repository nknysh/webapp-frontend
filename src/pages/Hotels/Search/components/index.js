export { default as HotelsGrid } from './HotelsGrid';
export { default as HotelSidebar } from './HotelSidebar';
