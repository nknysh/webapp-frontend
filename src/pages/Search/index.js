// Libraries
import React from 'react';
import qs from 'qs';
import { withRouter } from 'react-router-dom';

// App
import { colors } from 'styles';

// Components
import {
  Calendar,
  CoverPhoto,
  Dropdown,
  Form,
  Header,
  NumberInput,
  RadioButton,
  SelectInput,
  Styled,
} from 'components';

const Container = Styled.View.extend`
`;

const Fill = Styled.View.extend`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
`;

const Content = Styled.View.extend`
  width: 1160px;
  align-items: center;
  align-self: center;
  margin-vertical: 200px;
`;

const SearchContainer = Styled.View.extend`
  padding: 20px;
  background-color: ${colors.gray14};
`;

const Fields = Styled.View.extend`
  flex-direction: row;
  align-items: center;
`;

const Field = Styled.View.extend`
  margin-right: 20px;
`;

const Input = Styled.TextInput.H8.extend`
`;

const Options = Styled.View.extend`
`;

const SubmitButton = Styled.Button.extend`
  width: 150px;
`;

const SubmitText = Styled.H7.extend`
  color: ${colors.white16};
`;

const Search = ({ history }) => (
  <Container>
    <Header />
    <Fill>
      <CoverPhoto />
    </Fill>
    <Content>
      <SearchContainer>
        <Form
          initialValues={{
            country: '',
            checkIn: '',
            checkOut: '',
            numRooms: 1,
            numAdults: 0,
            numChildren: 0,
            numInfants: 0,
            honeymooners: false,

            // TODO(mark): Split into checkIn and checkOut.
            dates: '',
          }}
          onSubmit={values => {
            const query = qs.stringify(values);
            history.push(`/hotels?${query}`);
          }}
        >
          {({ values, handleChange, handleBlur, handleSubmit, setFieldValue, setFieldTouched }) => (
            <form>
              <Fields>
                <Field>
                  <SelectInput
                    name="country"
                    placeholder="DESTINATION OR RESORT"
                    value={values.country}
                    options={[{ value: 'maldives', label: 'MALDIVES' }, { value: 'seychelles', label: 'SEYCHELLES' }]}
                    onChange={setFieldValue}
                    onBlur={setFieldTouched}
                    style={{ width: 200 }}
                    Input={Input}
                  />
                </Field>
                <Calendar
                  name="dates"
                  placeholder="DATES"
                  value={values.dates}
                  onChange={handleChange}
                  onBlur={handleBlur}
                  style={{ width: 250 }}
                />
                <Field>
                  <Dropdown placeholder="ACCOMODATIONS" style={{ width: 175 }}>
                    <Options>
                      <NumberInput
                        name="numRooms"
                        placeholder="ROOMS"
                        value={values.numRooms}
                        onChange={setFieldValue}
                        onBlur={setFieldTouched}
                      />
                      <NumberInput
                        name="numAdults"
                        placeholder="ADULTS"
                        value={values.numAdults}
                        onChange={setFieldValue}
                        onBlur={setFieldTouched}
                      />
                      <NumberInput
                        name="numChildren"
                        placeholder="CHILDREN (2-12)"
                        value={values.numChildren}
                        onChange={setFieldValue}
                        onBlur={setFieldTouched}
                      />
                      <NumberInput
                        name="numInfants"
                        placeholder="INFANTS (0-2)"
                        value={values.numInfants}
                        onChange={setFieldValue}
                        onBlur={setFieldTouched}
                      />
                    </Options>
                  </Dropdown>
                </Field>
                <Field>
                  <RadioButton
                    name="honeymooners"
                    placeholder="HONEYMOONERS"
                    value={values.honeymooners}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Field>
                <SubmitButton onPress={handleSubmit}>
                  <SubmitText>SEARCH</SubmitText>
                </SubmitButton>
              </Fields>
            </form>
          )}
        </Form>
      </SearchContainer>
    </Content>
  </Container>
);

export default withRouter(Search);
