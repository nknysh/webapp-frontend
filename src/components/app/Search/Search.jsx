import React, { Fragment } from 'react';
import { __, path, lensProp, defaultTo, set, view, pipe, prop } from 'ramda';
import debounce from 'lodash.debounce';

import uiConfig from 'config/ui';
import { DatePicker, Checkbox } from 'components/elements';
import LodgingSelect from 'components/app/LodgingSelect';
import { isLoading } from 'store/common';

import { propTypes, defaultProps } from './Search.props';
import { SearchBarButton, SearchBarIndexSearch, SearchBarSection } from './Search.styles';

const destinationLens = lensProp('destination');
const datesLens = lensProp('dates');
const lodgingLens = lensProp('lodging');
const honeymoonersLens = lensProp('suitableForHoneymooners');

export const Search = ({
  indexes,
  searchStatus,
  searchQuery,
  onChange,
  onSearch,
  indexSelectors,
  onSubmit,
  showSubmit,
  vertical,
}) => {
  const updateSearchQuery = set(__, __, searchQuery);
  const getSearchQueryData = view(__, searchQuery);

  const setSelectedDatesToSearchQuery = pipe(
    updateSearchQuery(datesLens),
    onChange
  );

  const setLodgingsToSearchQuery = pipe(
    updateSearchQuery(lodgingLens),
    onChange
  );

  const setHoneymoonersToSearchQuery = pipe(
    updateSearchQuery(honeymoonersLens),
    onChange
  );

  const onIndexSelect = pipe(
    updateSearchQuery(destinationLens),
    onChange
  );

  return (
    <Fragment>
      <SearchBarSection data-vertical={vertical}>
        <SearchBarIndexSearch
          indexes={indexes}
          isLoading={isLoading(searchStatus)}
          label={path(['labels', 'search'], uiConfig)}
          limit={5}
          onChange={debounce(onSearch, 600)}
          onClick={onIndexSelect}
          openOnFocus={false}
          placeholder={path(['placeholders', 'search'], uiConfig)}
          selectors={indexSelectors}
          value={prop('value', getSearchQueryData(destinationLens))}
        />
      </SearchBarSection>
      <SearchBarSection data-vertical={vertical} data-large={true}>
        <DatePicker
          label={path(['labels', 'dates'], uiConfig)}
          onSelected={setSelectedDatesToSearchQuery}
          selectedValues={getSearchQueryData(datesLens)}
        />
      </SearchBarSection>
      <SearchBarSection data-vertical={vertical}>
        <LodgingSelect
          label={path(['labels', 'lodging'], uiConfig)}
          onSelected={setLodgingsToSearchQuery}
          selectedValues={getSearchQueryData(lodgingLens)}
        />
      </SearchBarSection>
      <SearchBarSection data-vertical={vertical} data-constrain={true}>
        <Checkbox
          label={path(['labels', 'honeymooners'], uiConfig)}
          onChange={(e, checked) => setHoneymoonersToSearchQuery(checked)}
          checked={defaultTo(false, getSearchQueryData(honeymoonersLens))}
        />
      </SearchBarSection>

      {showSubmit && (
        <SearchBarSection data-vertical={vertical}>
          <SearchBarButton onClick={onSubmit}>{path(['buttons', 'search'], uiConfig)}</SearchBarButton>
        </SearchBarSection>
      )}
    </Fragment>
  );
};

Search.propTypes = propTypes;
Search.defaultProps = defaultProps;

export default Search;
