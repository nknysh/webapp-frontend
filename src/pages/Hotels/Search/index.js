// Libraries
import React from 'react';
import { connect } from 'react-redux';

// App
import { fetchHotels } from 'store/modules/hotels/actions';
import { getHotels } from 'store/modules/hotels/selectors';
import { colors } from 'styles';

// Components
import { Header, Request, Styled } from 'components';
import { HotelSidebar, HotelsGrid } from './components';

const Container = Styled.View.extend`
`;

const Content = Styled.View.extend`
  align-self: center;
  width: 1280px;
  border-top-width: 1px;
  border-top-style: solid;
  border-top-color: ${colors.gray14};
  padding-top: 80px;
`;

const Row = Styled.View.extend`
  flex-direction: row;
`;

const Main = Styled.View.extend`
  flex: 1;
  margin-left: 30px;
`;

const Title = Styled.H4.extend`
  margin-bottom: 30px;
`;

const HotelsSearch = ({ searchHotels }) => (
  <Container>
    <Header />
    <Content>
      <Request getState={state => ({ hotels: getHotels(state) })} onRequest={searchHotels}>
        {({ hotels }) => (
          <Row>
            <HotelSidebar />
            <Main>
              <Title>{`${hotels.length} properties found`}</Title>
              <HotelsGrid hotels={hotels} />
            </Main>
          </Row>
        )}
      </Request>
    </Content>
  </Container>
);

// TODO(mark): Change fetchHotels --> searchHotels once we have derby soft keys.
export default connect(
  undefined,
  { searchHotels: fetchHotels }
)(HotelsSearch);
