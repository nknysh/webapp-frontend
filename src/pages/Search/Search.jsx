import React from 'react';

import { Markdown } from 'components';

import heroData from 'config/data/search/hero.md';
import heroImage from 'config/data/search/assets/hero-image.png';

import { StyledSearch, SearchHero } from './Search.styles';

export const Search = () => {
  return (
    <StyledSearch>
      <SearchHero mask media={{ image: heroImage }}>
        <Markdown>{heroData}</Markdown>
      </SearchHero>
    </StyledSearch>
  );
};

export default Search;
