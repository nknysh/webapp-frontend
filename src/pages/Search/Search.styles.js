import styled from 'styled-components';

import { BackButton, Container } from 'components';

import theme from 'styles/theme';
import breakpoints from 'styles/breakpoints';

export const StyledSearch = styled(Container)`
  width: 100%;

  ${breakpoints.tablet`
    padding: ${theme.gutter * 6}px 0;
  `}
`;

export const Columns = styled.div`
  ${breakpoints.tablet`
        display: flex;
    `}
`;

export const Column = styled.div`
  ${breakpoints.tablet`
        flex: 1;
    `}
`;

export const ColumnLeft = styled(Column)`
  ${breakpoints.tablet`
        flex-width: 27%;
        min-width: 27%;
        padding-right: ${theme.gutter * 3}px;
    `}
`;

export const ColumnRight = styled(Column)`
  ${breakpoints.tablet`
        flex-width: 73%;
        min-width: 73%;

    `}
`;

export const Back = styled(BackButton)`
  margin-bottom: ${theme.gutter * 6}px;
`;
