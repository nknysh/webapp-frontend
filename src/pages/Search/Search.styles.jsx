import styled from 'styled-components';

import { Hero } from 'components';

export const StyledSearch = styled.div`
  width: 100%;
`;

export const SearchHero = styled(Hero)`
  > * {
    max-width: 685px;
  }
`;
