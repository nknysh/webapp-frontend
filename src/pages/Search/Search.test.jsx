import React from 'react';

import { Search } from './Search';

const getComponent = props => shallow(<Search match={{ params: {} }} {...props} />);

describe('<Search />', () => {
  describe('render', () => {
    it('matches snapshot', () => {
      expect(getComponent()).toMatchSnapshot();
    });
  });
});
