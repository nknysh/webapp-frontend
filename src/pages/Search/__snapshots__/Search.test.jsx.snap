// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> render matches snapshot 1`] = `<Connect(SearchContainer) />`;
